### PR TITLE
cpuallocator: implement package prioritization

### PR DESCRIPTION
**Pull request by marquiz**
Try allocate all cpus from a single package, if possible. For each
allocation request calculate priority order between cpu packages, based
on requested number of cpus and the cpu priority preference.

### CODE REVIEW

**Comment by kad** [comment]
@marquiz Package is not enough. we need to check whole hierarchy: cluster->die->package

**Comment by marquiz** [comment]
> @marquiz Package is not enough. we need to check whole hierarchy: cluster->die->package

I know but I would leave that as a future improvement. We can discuss that in #191.

This PR is supposed to fix the greatest shortcoming in the current cpuallocator.

**Comment by askervin** [comment]
This change breaks a couple of e2e tests. I'm verifying if the new allocation order makes sense in these cases. For instance
`policies.test-suite/podpools/n4c16/test01-basic-placement`

(Thanks for very good debug prints. ;) )

File: pkg/cpuallocator/allocator.go [hunk 1]
```diff
@@ -343,6 +354,97 @@ func (a *allocatorHelper) allocate() cpuset.CPUSet {
...
+	pkgs := a.sys.PackageIDs()
+	offline := a.sys.Offlined()
+
+	pkgAvail := make(map[sysfs.ID]int, len(pkgs))
+	pkgPrioAvail := make(map[sysfs.ID][NumCPUPriorities]int, len(pkgs))
+	for _, pkg := range pkgs {
+		avail := a.topology.pkg[pkg].Difference(offline).Intersection(a.from)
+		prioAvail := [NumCPUPriorities]int{}
+		for prio := PriorityHigh; prio < NumCPUPriorities; prio++ {
+			prioAvail[prio] = avail.Intersection(a.topology.cpuPriorities[prio]).Size()
```
Comment (askervin) [inline pkg/cpuallocator/allocator.go:369]: On e2e test VM (topology n4c16), I'm getting `a.topology.cpuPriorities[prio] == [empty, empty, empty]`. Therefore `iTotal` and `jTotal` are always zero. Even in the "stealing" case. As a result, this function ends up sorting packages based on the number of available cores (line 434).

Yet this still results in very nice CPU allocation for instance in the podpools e2e tests, I'd like to make sure that happens also in the case when not ending up (unintentionally?) to this fallback branch.

**Comment by askervin** [comment]
@marquiz, suggesting changes/fixes.

Make sure `a.topology.cpuPriorities` is non-empty. It seems like it would be empty on any platform without this:
```
diff --git a/pkg/cpuallocator/allocator.go b/pkg/cpuallocator/allocator.go
index 4f9b74e0..daf71b4a 100644
--- a/pkg/cpuallocator/allocator.go
+++ b/pkg/cpuallocator/allocator.go
@@ -527,6 +527,7 @@ func (c *topologyCache) discoverCPUPriorities(sys sysfs.System) {
                        prio[p] = prio[p].Union(cset)
                }
        }
+       c.cpuPriorities = prio
 }
```

If the platform doesn't have cpufreq data, do you think all CPUs could be considered of Normal priority? Like adding following to `discoverSstCPUPriority()` (near `// Finally, determine priority of each CPU`):
```
if len(eppList) == 0 {
        if cpus, ok := epps[sysfs.EPPUnknown]; ok {
                prios[PriorityNormal] = cpus
        }
}
```

(Thanks to @klihub for spotting this!)

**Pull request closed without merging**